fvm/src/call_manager/default.rs

```diff
@@ -412,6 +416,198 @@ where
         Ok(())
     }
 
+    fn upgrade_actor<K: Kernel<CallManager = Self>>(
+        &mut self,
+        actor_id: ActorID,
+        new_code_cid: Cid,
+        params: Option<Block>,
+    ) -> Result<InvocationResult> {
+        let origin = self.origin;
+
+        let state = self
+            .state_tree_mut()
+            .get_actor(actor_id)?
+            .ok_or_else(|| syscall_error!(NotFound; "actor not found: {}", actor_id))?;
+
+        // store the code cid of the calling actor before running the upgrade endpoint
+        // in case it was changed (which could happen if the target upgrade endpoint
+        // sent a message to this actor which in turn called upgrade)
+        let code_id_before = state.code;
+
+        let result = self.upgrade_actor_inner::<K>(actor_id, new_code_cid, params)?;
+
+        if result.exit_code == ExitCode::OK {
+            // after running the upgrade, our code cid must not have changed
+            let code_id_after = self
+                .state_tree_mut()
+                .get_actor(actor_id)?
+                .ok_or_else(|| syscall_error!(NotFound; "actor not found: {}", actor_id))?
+                .code;
+            if code_id_before != code_id_after {
+                return Err(syscall_error!(Forbidden; "re-entrant upgrade detected").into());
+            }
+
+            self.state_tree_mut().set_actor(
+                origin,
+                ActorState::new(
+                    new_code_cid,
+                    state.state,
+                    state.balance,
+                    state.sequence,
+                    None,
+                ),
+            );
+        }
+
+        Ok(result)
+    }
+
+    fn upgrade_actor_inner<K: Kernel<CallManager = Self>>(
```

Comment:
Please deduplicate this with the "send" logic, it'll make it much easier to maintain and review.

Reply:
I.e., ideally we just have a single (internal) `call_actor` method on the call manager that takes an `Entrypoint`. We can either:

1. Use something like a visitor pattern (kind of complex but pluggable).
2. Make `Entrypoint` a "dumb" enum (like in https://github.com/filecoin-project/builtin-actors/pull/1339). This'll require a _little_ more logic in the `call_actor` method but shouldn't be too complex.

Externally, I think we'd still have to have both an upgrade and a send entrypoint.

Reply:
Also, see my comment on setting the code CID before upgrading. That should make this simpler.